Add c_module() manifest function for user C modules

andrewleech opened the pull request:
### Summary

This adds a `c_module()` function to the manifest system so projects can specify user C module directories directly in manifest.py files.

The `USER_C_MODULES` approach requires all C modules to exist within a single parent directory that gets scanned for modules. The new `c_module()` function provides two key benefits: it consolidates both Python and C modules in a single manifest location, and it allows each C module to be located anywhere in the codebase rather than requiring them all in one flat parent directory.

Example usage:
```python
c_module("$(MPY_DIR)/examples/usercmodule/cexample")
c_module("$(BOARD_DIR)/drivers/sensor")
```

The implementation adds early manifest processing during build configuration in both Make and CMake build systems. C module paths are extracted via `makemanifest.py --list-c-modules` and appended to the existing `USER_C_MODULES` list, maintaining full backward compatibility. Multiple `c_module()` calls are supported, and the same `$(VAR)` path substitution works as with other manifest functions. Invalid paths fail the build immediately with clear error messages.

I consolidated all manifest processing into a new `py/manifest.mk` file that handles variables, C module extraction, and frozen content generation. This eliminated duplication between py.mk and mkrules.mk.

### Testing

Tested on unix (coverage variant), esp32, rp2, and stm32 ports with the example C modules. All ports build successfully and modules are available at runtime.

Binary equivalence testing confirmed that builds using `c_module()` produce identical binaries to those using `USER_C_MODULES` on the command line. Verified on rp2 (RPI_PICO_W), stm32 (PYBV11), and unix (coverage) - all sections match byte-for-byte on embedded targets, with unix showing only expected gcov metadata differences.

During rp2 testing I found and fixed a bug where `FROZEN_MANIFEST` overrides weren't restoring the user manifest before usermod.cmake inclusion, causing `c_module()` calls to be silently ignored. The fix ensures all three methods (c_module() in board manifest, FROZEN_MANIFEST override, USER_C_MODULES) work correctly.

CI now tests `c_module()` on unix, esp32, rp2, and stm32 ports. The existing stm32 pyboard CI job still uses `USER_C_MODULES` to verify backward compatibility.

### Trade-offs and Alternatives

No runtime impact - this is purely build-time functionality. Build time overhead is minimal since manifest parsing was already happening for frozen modules.

One intentional behavioral difference: `USER_C_MODULES` scans a directory and includes all modules with micropython.mk/micropython.cmake files, while `c_module()` requires explicit listing of each module directory. This gives precise control over which modules are included rather than pulling in everything from a directory scan.

andrewleech commented:
Did a couple of small cleanups on top of this:

The new manifest processing logic added to `py/usermod.cmake` has been extracted out into its own `py/manifest.cmake`, parallel to `py/manifest.mk`. `usermod.cmake` just includes it now, so the file that ports reference stays unchanged. It was a bit odd having manifest-processing logic buried inside usermod.cmake with no obvious way to find it.

The `py/mkrules.mk` `include $(TOP)/py/manifest.mk` was also redundant - all ports include `py.mk` before `mkrules.mk`, and `manifest.mk` has an include guard (`MANIFEST_MK_INCLUDED`), so that second include was always a no-op. Removed it.

Quick sanity check on all four affected ports using the `manifest_test.py` files with `c_module()` entries - unix/coverage, PYBV11 (stm32), RPI_PICO_W (rp2), and ESP32_GENERIC (espressif idf container) - verified via `nm` on the ELF / cmake "Found User C Module(s)" output confirming all three modules linked in. On unix the modules also work at runtime: `cexample.add_ints(1, 2)` returns 3, `cppexample.cppfunc(3, 4)` returns `(7, 'hellocpp')`, and `example_package` is importable.

dpgeorge commented:
> The new manifest processing logic added to `py/usermod.cmake` has been extracted out into its own `py/manifest.cmake`,

> The `py/mkrules.mk` `include $(TOP)/py/manifest.mk` was also redundant - all ports include `py.mk` before `mkrules.mk`, and `manifest.mk` has an include guard (`MANIFEST_MK_INCLUDED`), so that second include was always a no-op. Removed it.

I don't see either of these changes in the latest PR here.  Maybe it wasn't pushed properly?

andrewleech committed:
tools/manifest: Add c_module() function for specifying user C modules.

Add a new c_module() function to the manifest API that lets a manifest
declare which user C module directories should be included in the build.
This pairs with the build system changes that pick up the resulting paths
and feed them into USER_C_MODULES.

The function validates the path exists, is a directory, and contains a
micropython.mk or micropython.cmake, and supports the usual \$(MPY_DIR),
\$(BOARD_DIR), \$(PORT_DIR), \$(MPY_LIB_DIR) substitutions. Unresolved
\$(VAR) tokens raise a clear error rather than failing later.

makemanifest.py gains --list-c-modules which prints one path per line so
the build system can pick them up (newline output handles paths that
contain whitespace on the cmake side).

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
examples/usercmodule: Fix subpackage cmake source filename.

The micropython.cmake referenced examplemodule.c but the source file is
named modexamplepackage.c. The subpackage example wasn't built via cmake
before, so the typo went unnoticed.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
docs: Document c_module() manifest function.

Document the new c_module() function in the manifest reference, including
the FROZEN_MANIFEST prerequisite, additive interaction with USER_C_MODULES,
and the make-side whitespace limitation. Add a corresponding section to
the cmodules developer guide so the feature is discoverable from both
entry points.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
tests: Add shared manifest_c_module test for c_module() CI.

Provides tests/tools/manifest_c_module.py, a shared frozen manifest that
pulls in the example user C modules via c_module() and includes the
port's default board manifest via $(PORT_DIR). Wires it up in ci.sh so
rp2 (RPI_PICO2), stm32 (PYBV11) and esp32 (SPIRAM variant) each exercise
the c_module() manifest path alongside the legacy USER_C_MODULES=
command-line variable.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
unix: Migrate coverage variant to use c_module() in manifest.

The unix coverage variant set USER_C_MODULES from mpconfigvariant.mk to
build the example user C modules into the test binary. Move that setting
into the coverage manifest as a c_module() call to exercise the new
manifest API end-to-end as part of the standard coverage test build.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
py: Add build system support for c_module() in manifests.

When MICROPY_FROZEN_MANIFEST is set, the build system runs makemanifest.py
with --list-c-modules to extract C module paths from the manifest and
appends them to USER_C_MODULES so they go through the same code path as
user modules specified on the command line.

The Make side lives in py/manifest.mk (included from py/py.mk) and the
CMake side lives in py/manifest.cmake (included from py/usermod.cmake).
Both share defaulting for the MICROPY_MANIFEST_* substitution variables
so a manifest can reference $(MPY_DIR) etc. consistently on either side.
The make-side extraction is gated on the build target needing it and on
micropython-lib being initialised, and surfaces makemanifest.py errors
via .SHELLSTATUS. The cmake side uses FATAL_ERROR to match.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
rp2: Resolve FROZEN_MANIFEST before usermod.cmake and quote for cmake.

Move the MICROPY_USER_FROZEN_MANIFEST resolution to run right after the
board config (rather than near the end of CMakeLists.txt) so it's applied
before py/usermod.cmake is included; py/manifest.cmake (invoked from
usermod.cmake) needs the final MICROPY_FROZEN_MANIFEST value to extract
c_module() paths. Also quote FROZEN_MANIFEST when forwarding it from the
Makefile to cmake so paths containing whitespace round-trip correctly.

Signed-off-by: Andrew Leech <andrew@alelec.net>

andrewleech committed:
esp32: Quote FROZEN_MANIFEST when forwarding it to idf.py.

Quote FROZEN_MANIFEST when passing it through IDFPY_FLAGS so paths
containing whitespace round-trip correctly to the cmake configure step;
py/manifest.cmake (auto-included via py/usermod.cmake) needs the
resolved value to extract c_module() paths from the manifest.

Signed-off-by: Andrew Leech <andrew@alelec.net>

dpgeorge commented:
Just a note that CI is failing here.

projectgus commented:
@andrewleech Feel free to request a review when you get a chance to look at this again.

andrewleech commented:
> @andrewleech Feel free to request a review when you get a chance to look at this again.

Thanks @projectgus - this should be in a good state now.

projectgus commented on tools/ci.sh:
```diff
@@ -254,14 +254,19 @@ function ci_esp32_build_common {
+    # build dir from the ESP32_GENERIC build above, so no cached cmake
+    # configuration combines across the two.
+    make ${MAKEOPTS} -C ports/esp32 BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM \
+        FROZEN_MANIFEST="$(pwd)/ports/esp32/boards/manifest_cmod_test.py"
```
For these CI tests, is there anything we can do to verify that the build not only succeeded but also included the specified C modules? Maybe grep for a symbol?

dpgeorge replied:
The unix coverage build includes cexample, cppexample and subpackage via this new `c_module()` mechanism.  And it then runs tests importing these, as well as printing `help('modules')`.  Those tests will fail if the C modules were not included in the build.

So, I think this is already covered.

projectgus reviewed (commented):
Hi @andrewleech, thanks for the ping. I haven't tested this out but the implementation looks sound - as @Gadgetoid also said last year - this seems like a very sensible approach.

One question about CI additions, but otherwise LGTM!

Gadgetoid commented:
As a big user of C modules I *really* ought to have given this a go and a conclusive 👍, sorry.

Currently looking into what this means for our boards and how far I can adapt. I suspect I'll still end up using CMake for some things, since I've a habit of needing to apply configuration tweaks, optimisations or sometimes even per-file settings to individual modules.

I even have stuff like this:

```cmake
set(PV_DUAL_CORE ON)
find_package(PICOVECTOR_MICROPYTHON CONFIG REQUIRED)

# Build picovector for Tufty 2350
target_compile_definitions(usermod_picovector INTERFACE TUFTY=1)
```

However taking a look at a more conventional and (popular?) extension for MicroPython - ulab - indicates that in-tree config files would be the canonical answer:

https://github.com/v923z/micropython-ulab/blob/2b6e0cfaa0d095e497e5a7599e0dbe3d09614ec2/code/ulab.h#L28-L32

However `ulab.h` does not include MicroPython's config as a means to access a custom config path outside of an explicit build arg. Perhaps it should.

Either here, or perhaps as an addendum in another PR, some documented canonical method of getting custom config into C modules might be appropriate.

(I don't think this otherwise materially affects this PR, or how `c_module` should work because putting module config anywhere but .h, Make or CMake will inevitably give everyone involved a headache.)

dpgeorge commented on tests/tools/test_manifestfile.py:
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
```
Maybe this test can be moved to `tests/tools/test_manifestfile.py`?  Then at least all the tests are in one location (can eventually move mpremote tests there as well).

andrewleech replied:
Thanks, I'd meant to ask for your advice on where to put this new kind of test script! I wasn't happy where it was either. A tests/tools folder makes perfect sense - done.

dpgeorge commented on ports/rp2/boards/manifest_test.py:
```diff
@@ -0,0 +1,9 @@
+# Manifest for testing the c_module() build feature.
```
This file is equivalent to `ports/stm32/boards/manifest_test.py`, and also `ports/esp32/boards/manifest_cmod_test.py` (up to change in comments).  So I suggest consolidating them all into one file that all ports can use, and put that somewhere in `tests/`.

andrewleech replied:
Good catch, yes I've consolidated and moved that too.

dpgeorge commented on tools/ci.sh:
```diff
@@ -254,14 +254,19 @@ function ci_esp32_build_common {
-    make ${MAKEOPTS} -C ports/esp32 BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM
+    # Exercise the c_module() codepath on the SPIRAM variant; a separate
+    # build dir from the ESP32_GENERIC build above, so no cached cmake
+    # configuration combines across the two.
```
I don't think the comment about "separate build dir" is needed.  That property is already necessary for all the other builds to succeed here.  Suggest the comment just be:
```
# Exercise the c_module() codepath on the SPIRAM variant.
```

dpgeorge commented on tools/ci.sh:
```diff
@@ -500,9 +490,12 @@ function ci_rp2_build {
     make ${MAKEOPTS} -C ports/rp2 BOARD=RPI_PICO2 submodules
-    make ${MAKEOPTS} -C ports/rp2 BOARD=RPI_PICO2
+    # Exercise c_module() on RPI_PICO2; a separate build dir from the
+    # RPI_PICO_W build above, so no cached cmake configuration combines.
```
As above, the comment about separate build dir is IMO adding more noise than information.  Suggest just:
```
# Exercise c_module() on RPI_PICO2.
```

dpgeorge commented on tools/manifestfile.py:
```diff
@@ -221,11 +229,16 @@ def _manifest_globals(self, kwargs):
+            # c_module() only has an effect in FREEZE / LIST_C_MODULES; in
+            # COMPILE / PYPROJECT modes it's a silent no-op so a manifest can be
+            # evaluated in any mode without raising NameError. Matches the way
+            # package()/module() behave in LIST_C_MODULES.
```
This comment can be removed, In my mind it adds more noise than information.  None of the other lines here have such documentation comments, yet they have similarly complex behaviour.  Furthermore, this behaviour is documented at the start of the `self.c_module` method, and it's much clearer there what is happening.

dpgeorge reviewed (approved):
Thanks @andrewleech for working on this.  It's a really great feature!  For a while now I did want to improve the situation with specifying USER_C_MODULES but what you did here is much better than any idea I had.  A great UI improvement.

Just a few minor comments about overly-verbose LLM commenting 😄  (In general I find that LLMs add comments about their reasoning/thought-process for a change, but after all the dust settles those comments really ought to be removed.)

andrewleech commented:
Thanks @dpgeorge yes I agree about the comments - they're usually excessive on first pass in agent development - I think those ones snuck back in during some of the more recent updates on this branch. 

I've updated them now thanks.

dpgeorge commented:
Thanks for updating.  Now merged!!
